### PR TITLE
deduplicate field access logic

### PR DESCRIPTION
**Pull request by oli-obk**
no regressions on rustc test suite detected

We treat arrays like structures with homogeneous fields.

### CODE REVIEW

File: src/lvalue.rs [hunk 1]
```diff
@@ -195,6 +195,19 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
 
+            // We treat arrays + fixed sized indexing like field accesses
+            Array { .. } => {
+                let field = field_index as u64;
+                let elem_size = match base_ty.sty {
+                    ty::TyArray(elem_ty, n) => {
+                        assert!(field < n as u64);
+                        self.type_size(elem_ty)?.expect("array elements are sized") as u64
+                    },
+                    _ => bug!("tried to access field of non-array type {:?}", base_ty),
```
Comment (bjorn3) [inline src/lvalue.rs]: This text can be confusing.

**Comment by oli-obk** [comment]
Alternatively we could skip the `Zst` primval entirely and use `Undef`, since there's nothing to be done with a zst anyway... Reading and writing it should be a noop.

File: src/error.rs [hunk 2]
```diff
@@ -57,6 +58,10 @@ pub enum EvalError<'tcx> {
     ExpectedDropGlue(Function<'tcx>),
     ManuallyCalledDropGlue,
     Panic,
+    NonZstToZstCast(PrimVal),
+    ZstToNonZstCast(Ty<'tcx>),
+    ZstToPtrComparison,
+    ReadFromZst,
```
Comment (solson) [inline src/error.rs]: All these new error cases make `PrimVal::Zst` seem less clean than the previous state.

**Comment by oli-obk** [comment]
Removed zsts + tiny rustup.

**Pull request merged**